### DISCUSSION TITLE
Order is not set but extended when ordering a field

### DISCUSSION THREAD

**Discussion by MMore** [woylie/flop_phoenix#87] [Ideas]
Hi Mathias,

I've just noticed that the order is not _set_ but actually _extended_ (via `push_order/2`). So when I have two fields A and B I can order by, I can do the the following:

1. Click on A and order by A
2. Click on B and order by B
3. Click on A and order by A, then click on B and order by A and B.
4. Click on B and order by B, then click on A and order by B and A.

But it's not possible to "unselect" an order criteria when it's enabled. Or just order by another field when one field order is already set. Would be nice to adjust that. What do you think?

However, I guess it would be possible to workaround that when handling single events.

Best
Mathias

**Comment by woylie** [comment]
I wonder whether that's really an issue. Say I have a table with IDs, names and ages. First I decide to sort by ages. Then I decide to sort by names. In the current implementation, at that point the table will be sorted by `name ASC age ASC`. If you say you want to sort _only_ by name, without having the age as the second sort column, you're effectively saying you are fine with pets that share the same name appearing in a random, non-deterministic order. But if you don't care about the order, why do you care that pets with the same name may explicitly _not_ be sorted by age?

There is another reason why you may not want to unsort columns. Most columns will not be unique. If you _only_ sort by non-unique columns, the order will be non-deterministic. Since the table will commonly be used in combination with pagination, this can result in unexpected results after going to the next or previous page. To make the order deterministic, you can set a default order that includes a unique column. In the case of the pet table, you could set the default order to `name ASC id ASC`. Since the table component will only ever prepend fields and never remove them, the ID will always be part of the order clause, and the order will always be deterministic. If you _replace_ the order clause instead of prepending it, you won't have that guarantee.

**Reply by MMore** [reply]
Good point. I think the keyword is "prepend". So changing the first criteria for order is basically what I need. That's fine. Maybe the visual representation is confusing. For instance, when I am ordering for 3 columns it's not clear what the first order criteria is (which might be the most important one) because all of them get a direction icon. What about showing the sorting direction indicator for the first criteria only?

**Reply by woylie** [reply]
Only showing one sort indicator might make it unclear to the user that sorting by multiple columns is an option. But the table adds an aria attribute to the table header column with the primary order field, so you can hide all non-primary sort indicators or change the styling based on that with CSS:

```css
th:not([aria-sort]) .order-direction {
  opacity: 0.2;
}
```

**Reply by MMore** [reply]
I can just speak for me and I found it confusing. I don't know how other users consider that though. However your workaround works for me, thank you!
